### PR TITLE
Add "container_include_condition" option to config.datadog

### PR DESCRIPTION
**Pull request by same-id**
Add "container_include_condition" option to config.datadog

Before this commit we could only use:

 * container_include
 * container_exclude

To choose which containers will be monitored by datadog.
And container_include takes precedence over container_exclude.

This makes it very hard/impossible to select which containers should
be monitored in a fine-grained manner.

In addition, if container_include is set it might ignore pause containers
which are supplied by the agent when exclude_pause_container is set as
detailed in the following issue:

  https://github.com/DataDog/datadog-agent/issues/6599

This commit adds `container_include_condition`, which lets you use
conditional expressions to better specify which containers to monitor.

More information about CEL, the conditional expression language used:

  https://github.com/google/cel-spec
  https://github.com/google/cel-go

For example, you can use the following condition expressions in the
configuration:

```
name == 'eu_gcr' || name == 'private_jfrog'
name == 'secret-container-dd' && image == 'docker-dd-agent'
name.matches('secret-container-.*') && image == 'docker-dd-agent'
```

The `container_include_condition` option is compatible with the
`exclude_pause_container` option.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: releasenotes/notes/introduce-cel-condition-in-container-filters-a2d0e49028367a08.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+features:
+  - |
+    List new features here, or remove this section.
```
Comment (kayayarai) [inline releasenotes/notes/introduce-cel-condition-in-container-filters-a2d0e49028367a08.yaml:11]:
```suggestion
```

File: releasenotes/notes/introduce-cel-condition-in-container-filters-a2d0e49028367a08.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    This commit introduces container_include_condition which uses CEL to better select which containers should be monitored.
```
Comment (kayayarai) [inline releasenotes/notes/introduce-cel-condition-in-container-filters-a2d0e49028367a08.yaml:13]:
```suggestion
    Adds `container_include_condition`, which lets you use conditional expressions to better specify which containers to monitor.
```

File: releasenotes/notes/introduce-cel-condition-in-container-filters-a2d0e49028367a08.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+    More information about CEL:
```
Comment (kayayarai) [inline releasenotes/notes/introduce-cel-condition-in-container-filters-a2d0e49028367a08.yaml:15]:
```suggestion
    More information about CEL, the conditional expression language used:
```

File: releasenotes/notes/introduce-cel-condition-in-container-filters-a2d0e49028367a08.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+    https://github.com/google/cel-spec
+    https://github.com/google/cel-go
+
+    Now we can have the following condition expressions in the configuration:
```
Comment (kayayarai) [inline releasenotes/notes/introduce-cel-condition-in-container-filters-a2d0e49028367a08.yaml:20]:
```suggestion
    For example, you can use the following condition expressions in the configuration:
```

File: releasenotes/notes/introduce-cel-condition-in-container-filters-a2d0e49028367a08.yaml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+    name == 'eu_gcr' || name == 'private_jfrog'
+    name == 'secret-container-dd' && image == 'docker-dd-agent'
+    name.matches('secret-container-.*') && image == 'docker-dd-agent'
```
Comment (kayayarai) [inline releasenotes/notes/introduce-cel-condition-in-container-filters-a2d0e49028367a08.yaml:24]:
~~~suggestion
   ```
    name == 'eu_gcr' || name == 'private_jfrog'
    name == 'secret-container-dd' && image == 'docker-dd-agent'
    name.matches('secret-container-.*') && image == 'docker-dd-agent'
   ```
~~~

File: releasenotes/notes/introduce-cel-condition-in-container-filters-a2d0e49028367a08.yaml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+    In addition it is compatible with the exclude_pause_container option.
```
Comment (kayayarai) [inline releasenotes/notes/introduce-cel-condition-in-container-filters-a2d0e49028367a08.yaml:26]:
```suggestion
    The `container_include_condition` option is compatible with the `exclude_pause_container` option.
```
Maybe say a couple words about what "is compatible" means? Do you mean the two options don't interfere with each other, or that the settings for one affect the settings of the other? or?

**Review by kayayarai — CHANGES_REQUESTED** [review]

**Comment by same-id** [comment]
Done, and added more info regarding compatibility

**Review by kayayarai — APPROVED** [review]

**Comment by same-id** [comment]
Anything else on my side I need to do?

go.mod

has a conflicting change already since I pushed my latest changes, not sure I should fix the conflict if the PR will stay open for a while

Thanks!

**Comment by same-id** [comment]
gentle ping

**Comment by clamoriniere** [comment]
Hi @same-id 

Could you give us some context about why you need this functionality? on which container orchestrator solution are you deploying the containers? Do you plan to exclude more containers compared to what you want to monitor?

Our current approach is to keep the exclude capability simple, and we try to push this capability on the application deployment side (instead of integrating it into the agent configuration). We did recently this PR https://github.com/DataDog/datadog-agent/pull/16489 to allow container exclusion based on Kubernetes Pod annotations. 
We can extend it to container labels if needed.

Do you think it can answer your needs?

**Comment by same-id** [comment]
Also, looking at the code, I assume you can not include a container by an annotation, if it's excluded in the agent config.
So basically annotations are exclusion based only.

I'm not too glad about this choice since I would have preferred all the containers to be excluded by default and then I choose which to include, I guess this was the original point of using `containerInclude` in the agent config.

**Comment by same-id** [comment]
@clamoriniere

**Comment by zhuminyi** [comment]
Closing this stale pr.

**Pull request closed without merging**